lib/logger/lib/logger/utils.ex

```diff
@@ -109,129 +114,62 @@ defmodule Logger.Utils do
-    do: collect_inspect(t, args, data, %{opts | limit: limit, width: :infinity})
-
-  defp collect_cc(:done, [?w | t], [data | args], _used_format, _used_args, opts),
-    do: collect_inspect(t, args, data, %{opts | width: :infinity})
-
-  defp collect_cc(:done, [?P | t], [data, limit | args], _used_format, _used_args, opts),
-    do: collect_inspect(t, args, data, %{opts | limit: limit})
-
-  defp collect_cc(:done, [?p | t], [data | args], _used_format, _used_args, opts),
-    do: collect_inspect(t, args, data, opts)
-
-  defp collect_cc(:done, [h | t], args, used_format, used_args, _opts) do
-    {args, used_args} = collect_cc(h, args, used_args)
-    {t, args, [h | used_format], used_args}
-  end
-
-  defp collect_cc(?x, [a, prefix | args], used), do: {args, [prefix, a | used]}
-  defp collect_cc(?X, [a, prefix | args], used), do: {args, [prefix, a | used]}
-  defp collect_cc(?s, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?e, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?f, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?g, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?b, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?B, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?+, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?#, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?c, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?i, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?~, args, used), do: {args, used}
-  defp collect_cc(?n, args, used), do: {args, used}
-
-  defp collect_inspect(t, args, data, opts) do
-    data =
-      data
-      |> Inspect.Algebra.to_doc(opts)
-      |> Inspect.Algebra.format(opts.width)
-    {t, args, 'st~', [data]}
-  end
-
-  defp collect_value(current, [?* | t], [arg | args], used_format, used_args, opts, next)
-      when is_integer(arg) do
-    collect_cc(next, t, args, [?* | used_format], [arg | used_args],
-               put_value(opts, current, arg))
-  end
-
-  defp collect_value(current, [c | t], args, used_format, used_args, opts, next)
-      when is_integer(c) and c >= ?0 and c <= ?9 do
-    {t, c} = collect_value([c | t], [])
-    collect_cc(next, t, args, c ++ used_format, used_args,
-               put_value(opts, current, c |> :lists.reverse |> List.to_integer))
-  end
-
-  defp collect_value(_current, t, args, used_format, used_args, opts, next),
-    do: collect_cc(next, t, args, used_format, used_args, opts)
-
-  defp collect_value([c | t], buffer)
-    when is_integer(c) and c >= ?0 and c <= ?9,
-    do: collect_value(t, [c | buffer])
-
-  defp collect_value(other, buffer),
-    do: {other, buffer}
-
-  defp put_value(opts, key, value) do
-    if Map.has_key?(opts, key) do
-      Map.put(opts, key, value)
-    else
-      opts
-    end
+    format
+    |> :io_lib.scan_format(args)
+    |> handle_format_list(opts, [])
```

Comment:
handle_format_List is a map operation, maybe we should write it as a map?

```
|> Enum.map(&handle_format_spec(&1, &2))
```

And then make `handle_format_spec` returns the first argument as is if it is not a map.

Reply:
You're right! And `handle_format_spec` already returns the first argument if it is not a map. ;)